docs(core): edit and add docs for ng-template

dario-piotrowicz opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


Issue Number: https://github.com/angular/angular/issues/42678


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

dario-piotrowicz committed:
docs(core): edit and add docs for ng-template

mary-poppins commented:
You can preview 41584cd at https://pr42704-41584cd.ngbuilds.io/.

petebacondarwin commented on aio/content/special-elements/core/ng-template.md:
```diff
@@ -1,3 +1,33 @@
+
+- **have a context**
+
+    Inside the `<ng-template>` tags developers can reference variables present in the outer template in which the elements are defined, moreover `<ng-template>` elements can also have an associated context object containing variables which can be accessed via template (`let` and `as`) declarations.
```
Can we provide an example of how to do this?

dario-piotrowicz replied:
maybe, but this is already covered in the [`NgTemplateOutlet` docs](https://angular.io/api/common/NgTemplateOutlet)

maybe I could put an example using `createEmbeddedView`?

petebacondarwin commented on aio/content/special-elements/core/ng-template.md:
```diff
@@ -1,3 +1,33 @@
+
+    `<ng-template>` elements can be references using [standard template variables](/guide/template-reference-variables#how-angular-assigns-values-to-template-variables).
+
+    _This is for example how `<ng-template>` elements are used as [ngIf](/api/common/NgIf) else clauses._
```
```suggestion
    _This is how `<ng-template>` elements are used as [ngIf](/api/common/NgIf) else clauses._
```

petebacondarwin commented on aio/content/special-elements/core/ng-template.md:
```diff
@@ -1,3 +1,33 @@
+
+    _This is for example how `<ng-template>` elements are used as [ngIf](/api/common/NgIf) else clauses._
+
+    Such template variables can also be used to reference `<ng-template>` elements in the component's template when using [`ngTemplateOutlet`](/api/common/NgTemplateOutlet) directives.
```
Should we show an example of this too?

dario-piotrowicz replied:
again this is already presented in the `ngTemplateOutlet` docs and since they are referenced there directly it is very straightforward for visitors to just click the link and see the example

so I am not totally sure, I do can put a small example anyway if you feel it'd be good

please let me know what you think about it :slightly_smiling_face:

petebacondarwin commented on aio/content/special-elements/core/ng-template.md:
```diff
@@ -1,3 +1,33 @@
+    Such template variables can also be used to reference `<ng-template>` elements in the component's template when using [`ngTemplateOutlet`](/api/common/NgTemplateOutlet) directives.
+
+
+- **can be used to programmatically**
```
```suggestion
### Querying for templates
```

petebacondarwin commented on aio/content/special-elements/core/ng-template.md:
```diff
@@ -1,3 +1,33 @@
+
+- **can be used to programmatically**
+
+    A [Query](/api/core/Query) can find the `TemplateRef` associated to an `<ng-template>` element and then it can be used to directly create embedded views using the `ViewContainerRef`'s `createEmbeddedView` method.
```
I think it might be better to talk about ContentQuery and ViewQuery explicitly rather than the generic Query class... what do you think?

An example would also be very helpful here.

dario-piotrowicz replied:
About the generic `Query` I kind of just followed the [`templateRef` docs](https://angular.io/api/core/TemplateRef#description), I may amend it I'd imagine (would you also like me to chage the `templateRef` docs?)

About the example, yeah I actually had a quick look and at a first glance no such example does not seem to in the docs, so that would definitely be a good addition :slightly_smiling_face: 

but I am not 100% sure if it should be added here or in the `templateRef` docs, as it is kind of more related to that, what do you think?

petebacondarwin replied:
I think you are right that this doc should just reference the detail (including examples) in other docs.
If there is no example for this, perhaps an extra PR with such an example would be a good future task.

petebacondarwin reviewed (changes requested):
Great start - thanks for identifying all the use cases and relevant aspects of ng-template.
I left a lot of comments. Feel free to take what you think works.

petebacondarwin commented:
OK - so let's go without the examples for now. We will get the docs team (e.g. @TeriGlover) to go over it once you are happy.

dario-piotrowicz commented:
Thanks a lot @petebacondarwin yeah ok, I like the idea let's leave the examples for later
(we can add them to the issue and track the progress there)

I will get on with the changes and will let you know when I am happy with them :slightly_smiling_face:

dario-piotrowicz commented on aio/content/special-elements/core/ng-template.md:
```diff
@@ -1,3 +1,35 @@
+
+One of the main uses for `<ng-template>` is to hold template content that will be used by [Structural directives](/guide/structural-directives). Those directives can add and remove copies of the template content based on their own logic.
+
+> When using the [structural directive shorthand](/guide/structural-directives#structural-directive-shorthand), Angular creates an `<ng-template>` element behind the scenes.
```
I tired using `>` here and below.... I am not completely sure about the result...

petebacondarwin replied:
I think we should avoid block quotes. They are not used in any of our other docs.
If you want to call out this paragraph then you could use an [alert or callout](https://angular.io/guide/docs-style-guide#alerts-and-callouts).
```

dario-piotrowicz replied:
yeah I considered using alerts, but they seems to be too strong, this text is not that important

I think I'll just leave it as normal text

dario-piotrowicz commented:
@petebacondarwin I hope I am not bothering you too much, I made changes and I was hoping you could give it another glance? :sweat: 

I followed you comments but I changed something here and there, I hope it's fine :slightly_smiling_face:

mary-poppins commented:
You can preview 4ce7c48 at https://pr42704-4ce7c48.ngbuilds.io/.

petebacondarwin commented on aio/content/special-elements/core/ng-template.md:
```diff
@@ -1,3 +1,35 @@
+### TemplateRef
+
+`<ng-template>` elements are represented as instances of the `TemplateRef` class.
+This class exposes methods to make use of the defined template's content.
```
Actually i don't think one generally uses this object directly, right?
Instead what about saying:

```suggestion
Pass this object to `ViewContainerRef.createEmbeddedView()` in order to add copies of the template to the DOM.
```

dario-piotrowicz replied:
you where the one who suggested that change! :stuck_out_tongue_winking_eye: 

anyway I am not totally convinced in mentioning `createEmbeddedView` here, as it is already mentioned below, and also `TemplateRef`s have more broad uses, for example they are used by `ngTemplateOutlet`

so I will change the text, but it's ok with you I would try to keep it generic here :slightly_smiling_face:

petebacondarwin replied:
Sorry if I misled you with my suggestions. They are more a guide than a requirement.
As long as these are technically correct, we can leave the docs-team to decide upon stylistic and language improvements.

dario-piotrowicz replied:
nono no need to apologise I was kidding :sweat: 

(and I so much appreciate all your help :slightly_smiling_face: )

I'll think of something to put here :slightly_smiling_face:

dario-piotrowicz replied:
Actually I have been thinking and it is a bit redundant to mention that the `TemplateRef` gets used by `ngTemplateOutlet` as that is sort of implied below, so there is no point in mentioning that again here

Besides that and `createEmbeddedView` I am not really sure if there are other uses of `TemplateRef`s to be honest :cry: 

 So I am just going with what you suggested, though I wonder if mentioning twice `ViewContainerRef.createEmbeddedView()` will look ugly :thinking:

petebacondarwin commented on aio/content/special-elements/core/ng-template.md:
```diff
@@ -1,3 +1,35 @@
+
+### Querying
+
+A [Query](/api/core/Query) (such as `ViewChild`) can find the `TemplateRef` associated to an `<ng-template>` element so that it can be used programmatically. Generally to directly create embedded views via the `ViewContainerRef`'s `createEmbeddedView` method.
```
This last sentence doesn't make sense to me:

> Generally to directly create embedded views via the `ViewContainerRef`'s `createEmbeddedView` method.

Did you mean something like:

> For instance, to pass it to the `ViewContainerRef.createEmbeddedView()` method.

dario-piotrowicz replied:
Sorry if my wording is confusing, what I tried to convey is that the most common use of a queried `TemplateRef` is to create a view using `createEmbeddedView` and that it id done directly as opposed to other views which created indirectly (by Angular)

I'll amend this thanks :slightly_smiling_face:

dario-piotrowicz replied:
Sorry I just noticed that with `ViewContainerRef.createEmbeddedView()` we don't get the text to link to the method

Would the original version not be ok for you? 
```
`ViewContainerRef`'s `createEmbeddedView()` method
```

dario-piotrowicz replied:
Alternatively I could even go with: 
```
`ViewContainerRef`.`createEmbeddedView()`
```
which would look like this:
![Screenshot at 2021-07-01 19-02-56](https://user-images.githubusercontent.com/61631103/124170236-16aca300-da9f-11eb-9674-52a10966f901.png)

dario-piotrowicz replied:
or I can just copy what the TemplateRef docs do:
![Screenshot at 2021-07-01 19-08-44](https://user-images.githubusercontent.com/61631103/124170927-d568c300-da9f-11eb-8500-bf9f6926c0b4.png)

as in:
![Screenshot at 2021-07-01 19-11-28](https://user-images.githubusercontent.com/61631103/124171240-34c6d300-daa0-11eb-802c-2c9708f37a13.png)

petebacondarwin replied:
Gosh! You are right. We should actually fix the auto-code linker to handle that better. But in the meantime, feel free to choose a form that you feel works best for you. Perhaps copying the `TemplateRef` approach for consistency?

dario-piotrowicz replied:
ok yeah, I just copied `TemplateRef`'s approach then :slightly_smiling_face:

petebacondarwin commented on aio/content/special-elements/core/ng-template.md:
```diff
@@ -1,3 +1,35 @@
+
+### Context
+
+Inside the `<ng-template>` tags you can reference variables present in the surrounding outer template, additionally a context objects can be associated to `<ng-template>` elements, such object can contain variables which can be accessed within the element via template (`let` and `as`) declarations.
```
```suggestion
Inside the `<ng-template>` tags you can reference variables present in the surrounding outer template.
Additionally, a context object can be associated with `<ng-template>` elements.
Such an object contains variables which can be accessed from within the template contents via template (`let` and `as`) declarations.
```

petebacondarwin reviewed (commented):
Looking good. I made a few more suggestions.

mary-poppins commented:
You can preview 2e8d3fd at https://pr42704-2e8d3fd.ngbuilds.io/.

mary-poppins commented:
You can preview 228c7fe at https://pr42704-228c7fe.ngbuilds.io/.

dario-piotrowicz commented:
@petebacondarwin Thanks for all your help :slightly_smiling_face: 

I made the PR ready for review as I hope that now the docs seem good enough to be passed on to the docs team

Hope you are happy with them too :slightly_smiling_face:

petebacondarwin reviewed (approved):
This is a considerable improvement over the current minimal docs for ng-element. Thanks @dario-piotrowicz.

Let's see what @aikidave or @TeriGlover have to say about styling, although I am not sure if they are online right now. We might need to wait a little while to get their review.

dario-piotrowicz commented:
Thanks a lot @petebacondarwin 

If this goes well the other two special elements are next :stuck_out_tongue_winking_eye: 

(plus the examples we said I could add for the templates :slightly_smiling_face: )

jessicajaniuk reviewed (approved):
LGTM :cookie:

This is great! Thanks for this work!

mary-poppins commented:
You can preview 5579ea0 at https://pr42704-5579ea0.ngbuilds.io/.

TeriGlover commented on aio/content/special-elements/core/ng-template.md:
```diff
@@ -1,3 +1,36 @@
+### Context
+Inside the `<ng-template>` tags you can reference variables present in the surrounding outer template.
+Additionally, a context object can be associated with `<ng-template>` elements.
+Such an object contains variables which can be accessed from within the template contents via template (`let` and `as`) declarations.
```
```suggestion
Such an object contains variables that can be accessed from within the template contents via template (`let` and `as`) declarations.
```

TeriGlover reviewed (approved):
Copy edit complete; several comments. Otherwise, LGTM.

dario-piotrowicz commented on aio/content/special-elements/core/ng-template.md:
```diff
@@ -1,3 +1,36 @@
+Angular's `<ng-template>` element defines a template that is not rendered by default.
 
-With `<ng-template>`, you can render the content manually for full control over how the content displays.
+With `<ng-template>`, you can define template content that is only being rendered by Angular when you (directly or not) specifically instruct it to do so, allowing you to have full control over how and when the content is displayed.
```
Yeah I didn't see that before but that "specifically" part definitely needs improvements, I will look into it thanks! :slightly_smiling_face:

dario-piotrowicz replied:
About the "Further information available in the Usage Notes..."

That text is being set by dgeni in the docs generation as you can see here:
https://github.com/angular/angular/blob/9d58ebfcee7aef63b71bc133fc9f172d63fc9b14/aio/tools/transforms/templates/api/includes/description.html#L5

(and it is generated by the "@usages" below)

So I am not 100% sure if I can just change that as I am afraid that that is just out of scope here and also I am not sure how many docs actually use it, maybe I could investigate a bit and create a new PR for that change...

anyway I just wanted to say that it'd be a global change and is by no means restricted to this file

@petebacondarwin could you please advice on what to do here? :slightly_smiling_face:

dario-piotrowicz commented on aio/content/special-elements/core/ng-template.md:
```diff
@@ -1,3 +1,36 @@
+
+`<ng-template>` elements are represented as instances of the `TemplateRef` class.
+
+Pass this object to the `ViewContainerRef` method `createEmbeddedView()` in order to add copies of the template to the DOM.
```
yeah your version looks nicer and clearer thanks! :slightly_smiling_face:

dario-piotrowicz commented:
@TeriGlover Thanks a lot for the review :slightly_smiling_face: , I addressed all your comments, all besides the Usage Notes text, I can amend that text, but would love to hear from @petebacondarwin if he agrees with my changing it here :slightly_smiling_face:

mary-poppins commented:
You can preview f290f84 at https://pr42704-f290f84.ngbuilds.io/.

mary-poppins commented:
You can preview 68f1b49 at https://pr42704-68f1b49.ngbuilds.io/.

gkalpak reviewed (commented):
Drive-by review:
Please update the commit message to be `docs(core): ...`.
`docs(docs-infra): ...` means documentation about the documentation infrastructure (which is not what this PR is about :grin:).

dario-piotrowicz commented:
@gkalpak sorry about that, I've fixed it thanks! :slightly_smiling_face:

mary-poppins commented:
You can preview b9ea229 at https://pr42704-b9ea229.ngbuilds.io/.

aikithoughts reviewed (approved):
Thank you so much for this addition! LGTM!

Reviewed-for: global-docs-approvers
